core/src/main/java/org/apache/calcite/rex/RexExecutable.java

```diff
@@ -85,7 +85,7 @@ public void reduce(RexBuilder rexBuilder, List<RexNode> constExps,
         reducedValues.add(
             rexBuilder.makeLiteral(value.right, value.left.getType(), true));
       }
-    } catch (RuntimeException e) {
+    } catch (RuntimeException | AssertionError e) {
```

Comment:
I don't think we should ever catch AssertionError. Can't we avoid this?

Reply:
The ```rexBuilder.makeLiteral(value.right, value.left.getType(), true));``` can't create literal for non-standard types, such as GEOMETRY. Since the catch handles failed attempts maybe it's not so bad to add assertion error here. What do you think?

Reply:
Shouldn't we always keep the `rexBuilder.makeLiteral` works correctly ?

Reply:
@danny0405 is it possible to make literal for each type? For example, what to do here with user-defined types?

Reply:
I definitely won't add support for each type of literal, if you're not ok with ```AssertionError```, I can throw ```IllegalArgumentException``` instead from ```rexBuilder.makeLiteral(...)```.

Reply:
Technically speaking, `throw AssertionError` was added by Julian 7 years or so ago: https://github.com/apache/calcite/blob/571731b80a58eb095ebac7123285c375e7afff90/core/src/main/java/org/apache/calcite/rex/RexBuilder.java#L1487-L1488

I'm inlined that `IllegalArgumentException` works better there. @zabetak , @julianhyde , any thoughts?

In other words, I suppose that "AssertionError == something weird has happened", and "IllegalArgumentException == it worked as designed, but the argument value is wrong".

The sad thing altering the exception type might break the backward compatibility (e.g. third-party code might rely on `catch AssertionError` from `makeLiteral`).


So, to move this PR forward, I suggest: we leave `catch AssertionError` here, add a code comment that explains why `AssertionError` needs be here and add a TODO item to `makeLiteral` to refactor AssertionError into `IllegalArgument` or something like that.

Reply:
Whether to use `AssertionError` vs `IllegalArgumentException` depends on whether this is a private or public method. I had assumed that it was private, so if the value is not compatible with the type, it is a bug in Calcite.

I'd rather not add a TODO. Let's change it to `IllegalArgumentException`. Any code that is depending on `AssertionError` deserves to be broken. (Because only internal APIs throw `AssertionError`, ergo, you shouldn't be calling this API.)

By the way, in 2014 I changed the argument type from `Comparable` to `Object`. I think that was because I wanted to allow `Boolean`, and `Boolean` was not `Comparable` until about JDK 1.7. If we're deprecating things or changing APIs, we could change it back.

Reply:
Regarding `RexExecutable`. Why do you need to catch `AssertionError`? If `AssertionError` happens there is a bug in Calcite code. We should not catch that.

Reply:
Hi @julianhyde , here is what happens for ```spatial.iq``` file in ```CoreQuidemTest.java```: 

```
> java.lang.AssertionError: Was not expecting value 'GEOMETRY' for enumeration 'org.apache.calcite.sql.type.SqlTypeName' in this context
> 	at org.apache.calcite.util.Util.unexpected(Util.java:1785)
> 	at org.apache.calcite.rex.RexBuilder.makeLiteral(RexBuilder.java:1488)
> 	at org.apache.calcite.rex.RexBuilder.makeLiteral(RexBuilder.java:1371)
> 	at org.apache.calcite.rex.RexExecutable.reduce(RexExecutable.java:86)
> 	at org.apache.calcite.rex.RexExecutorImpl.reduce(RexExecutorImpl.java:130)
> 	at org.apache.calcite.rel.rules.ReduceExpressionsRule.reduceExpressionsInternal(ReduceExpressionsRule.java:695)
> 	at org.apache.calcite.rel.rules.ReduceExpressionsRule.reduceExpressions(ReduceExpressionsRule.java:616)
> 	at org.apache.calcite.rel.rules.ReduceExpressionsRule$FilterReduceExpressionsRule.onMatch(ReduceExpressionsRule.java:176)
```
The code tried to replace function calls like ```ST_MAKEPOINT(51.5:DECIMAL(3, 1), -0.12:DECIMAL(3, 2))``` with literal value.